libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -641,42 +642,24 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		break;
 	case FunctionType::Kind::Internal:
 	{
-		optional<FunctionDefinition const*> functionDef;
-		if (auto memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
-		{
-			solUnimplementedAssert(!functionType->bound(), "Internal calls to bound functions are not yet implemented for libraries and not allowed for contracts");
+		FunctionDefinition const* unresolvedFunctionDef = IRHelpers::referencedFunctionDeclaration(_functionCall.expression());
 
-			functionDef = dynamic_cast<FunctionDefinition const*>(memberAccess->annotation().referencedDeclaration);
-			if (functionDef.value() != nullptr)
-				solAssert(functionType->declaration() == *memberAccess->annotation().referencedDeclaration, "");
-			else
-			{
-				solAssert(dynamic_cast<VariableDeclaration const*>(memberAccess->annotation().referencedDeclaration), "");
-				solAssert(!functionType->hasDeclaration(), "");
-			}
-		}
-		else if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
-		{
-			solAssert(!functionType->bound(), "");
+		solUnimplementedAssert(!memberAccess || !functionType->bound(), "");
```

Comment:
These assertions are all rather difficult to read. Would it be better to use something like
```
if (a)
  solAssert(!b);
```
?

Reply:
On the other hand I like that each one is now self-contained instead of having multiple nesting levels. That's a bit less of a problem now that I introduced `referencedFunctionDeclaration()` but before that I was getting a bit lost in it because the three main branches (identifier vs member vs something else) had overlapping but not exactly identical checks.

Also, they're just extra sanity checks, not a part of the logic and if they're not compact they obscure the more important bits of code in my opinion.

I agree that they're harder to read though. Before I go on with this change, here's a side-by side comparison with the best nested version I could come up with:
``` c++
solUnimplementedAssert(!memberAccess || !functionType->bound(), "");
solAssert(memberAccess || !functionType->bound(), "");

solAssert(!memberAccess || !unresolvedFunctionDef || functionType->declaration() == *unresolvedFunctionDef, "");
solAssert(unresolvedFunctionDef || !functionType->hasDeclaration(), "");

// functionDef = ...

solAssert((functionDef == nullptr) == (unresolvedFunctionDef == nullptr), "");
solAssert(!functionDef || functionDef->isImplemented(), "");
```

``` c++
if (memberAccess)
	solAssert(!functionType->bound(), "");
else
	solUnimplementedAssert(!functionType->bound(), "");

if (unresolvedFunctionDef)
{
	if (!memberAccess)
		solAssert(functionType->declaration() == *unresolvedFunctionDef, "");
}
else
	solAssert(!functionType->hasDeclaration(), "");

// functionDef = ...

if (functionDef)
{
	solAssert(unresolvedFunctionDef, "");
	solAssert(functionDef->isImplemented(), "");
}
else
	solAssert(!unresolvedFunctionDef, "");
```

Should I change it?

Reply:
I changed it after all. The version with `if`s is not that bad when it's not mixed with logic and the compact one is indeed harder to read. I just noticed that the condition for `unresolvedFunctionDef && memberAccess` should really have been just `unresolvedFunctionDef`.

Reply:
Not so sure about either :)

Reply:
I tweaked it a bit more. Here's how it looks like now:

``` c++
if (functionDef)
{
	solAssert(memberAccess || identifier, "");
	solAssert(functionType->declaration() == *functionDef, "");

	if (identifier)
		functionDef = &functionDef->resolveVirtual(m_context.mostDerivedContract());

	solAssert(functionDef->isImplemented(), "");
}
else
	solAssert(!functionType->hasDeclaration(), "");

if (memberAccess)
	solUnimplementedAssert(!functionType->bound(), "");
else
	solAssert(!functionType->bound(), "");
```